Feature/update deployment (#61)
GenomicDataInfrastructure/starter-kit-storage-and-interfaces: README.md

| Component    | Description                                                                                                                                                                                                          |
|--------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| broker       | RabbitMQ based message broker, [SDA-MQ](https://github.com/neicnordic/sensitive-data-archive/tree/main/rabbitmq).                                                                                                    |
| database     | PostgreSQL database, [SDA-DB](https://github.com/neicnordic/sensitive-data-archive/tree/main/postgresql).                                                                                                            |
| storage      | S3 object store, demo uses Minio S3.                                                                                                                                                                                 |

> | auth         | OpenID Connect relaying party and authentication service, [SDA-auth](https://github.com/neicnordic/sensitive-data-archive/tree/main/sda-auth).                                                                       |

Reviewer: ```suggestion
| auth         | OpenID Connect relaying party and authentication service, [SDA-auth](https://github.com/neicnordic/sensitive-data-archive/tree/main/cmd/auth/auth.md).                                                                       |
```

> | s3inbox      | Proxy inbox to the S3 backend store, [SDA-S3Inbox](https://github.com/neicnordic/sensitive-data-archive/tree/main/sda).                                                                                              |

Reviewer: ```suggestion
| s3inbox      | Proxy inbox to the S3 backend store, [SDA-S3Inbox](https://github.com/neicnordic/sensitive-data-archive/tree/main/sda/cmd/s3inbox/s3inbox.md).                                                                                              |
```

Maybe this should be included in the SDA below.

Author: Hmm yeah makes sense, given its part of the cmd/sda, but its not mentioned in the [cmd/sda/sda.md](https://github.com/neicnordic/sensitive-data-archive/tree/main/sda/sda.md) atm, and would make sense to include `auth` there as well if its been moved there

| download     | Data out solution for downloading files from the SDA, [SDA-download](https://github.com/neicnordic/sensitive-data-archive/tree/main/sda-download).                                                                   |
| SDA-pipeline | The ingestion pipeline of the SDA, [SDA-pipeline](https://github.com/neicnordic/sensitive-data-archive/tree/main/sda). This comprises of the following core components: `ingest`, `verify`, `finalize` and `mapper`. |

> Detailed documentation on the `sda-pipeline` can be found at: [https://neic-sda.readthedocs.io/en/latest/services/pipeline)](https://neic-sda.readthedocs.io/en/latest/services/pipeline).

Reviewer 2: ```suggestion
Detailed documentation on the `sda-pipeline` can be found at: [https://neic-sda.readthedocs.io/en/latest/services/sda/)](https://neic-sda.readthedocs.io/en/latest/services/sda/).
```

The old link https://neic-sda.readthedocs.io/en/latest/services/pipeline is outdated.

Lastly, this repo includes a `docker-compose-demo.yml` file which deploys a standalone stack along with a demo of the sda services' functionality with test data. Details can be found [here](./README.md##Starting-the-stack-in-standalone-demo-mode).

### Adding TLS to internet facing services

Internet facing services such as `s3inbox`, `download` and `auth`, need to be secured via TLS certification. This can be most conveniently achieved by using [Let's Encrypt](https://letsencrypt.org/getting-started/) as Certificate Authority. Assuming shell access to your web host, a convenient way to set this up is through installing Certbot (or any other ACME client supported by Let's Encrypt). Detailed instructions on setting up Certbot for different systems can be found [here](https://certbot.eff.org/).

For an example of how to set up the `sda-pipeline` with TLS, see [TLS-example/README.md](config/TLS-example/README.md)

## Authentication for users with LS-AAI (mock or alive)

> To interact with SDA services, users need to provide [JSON Web Token](https://jwt.io/) (JWT) authorization. Ultimately, tokens can be fetched by [LS-AAI](https://lifescience-ri.eu/ls-login/) upon user login to an OpenID Connect (OIDC) relaying party (RP) service that is [registered with LS-AAI](https://spreg-legacy.aai.elixir-czech.org/). An example of such an RP service is the [sda-auth](https://github.com/neicnordic/sda-auth), which is included in the present stack.

Reviewer 2: I think this link https://spreg-legacy.aai.elixir-czech.org/ is outdated. Maybe https://services.aai.lifescience-ri.eu/spreg/ is the correct one. Is that correct @jbygdell ?

Also the link to `sda-auth` should be changed to https://github.com/neicnordic/sensitive-data-archive/tree/main/sda/cmd/auth

Reviewer: Yes, all mentions of `elixir-czech` should be converted to the new ones  from `lifescience-ri.eu`.
That means the URLs needs to be verified since they are not just a hostname change.

Author: Not sure I follow what this entails in terms of this PR, does the user of the starter kit need to know the OIDC URL themselves and repo should not state it?

```

It is possible to download the `s3cmd.conf` file from the `sda-auth` service as described in the [Authentication for users with LS-AAI (mock or alive)](#authentication-for-users-with-ls-aai-mock-or-alive) section above. However, do note that `s3cmd.conf` downloaded from this service lacks the section header `[default]` which needs to be added manually if one wishes to use the file directly with `s3cmd`.

For example, a `s3cmd.conf` file downloaded from `auth` after deploying the stack locally (with LS-AAI-mock as OIDC) would look like this:

```ini
access_key = jd123_lifescience-ri.eu
secret_key = jd123_lifescience-ri.eu
access_token=eyJraWQiOiJyc2ExIiwidH...
check_ssl_certificate = False
check_ssl_hostname = False
encoding = UTF-8
encrypt = False
guess_mime_type = True
host_base = localhost:8000
host_bucket = localhost:8000
human_readable_sizes = true
multipart_chunk_size_mb = 50
use_https = True
socket_timeout = 30
```

where the access token has been truncated for brevity. Please note that the option `use_https = True` is missing from the above file (therefore set implicitly to `False`) since the local deployment of the stack does not use TLS.

### The sda-cli tool

> Instead of the tools above, users are **encouraged** to use [`sda-cli`](https://github.com/NBISweden/sda-cli), which is a tool specifically developed to perform all common SDA user-related tasks in a convenient and unified manner. It is recommended to use precompiled executables for `sda-cli` which can be found at [https://github.com/NBISweden/sda-cli/releases](https://github.com/NBISweden/sda-cli/releases)

Reviewer 2: We probably need to specify a version of `sda-cli` to use.

Reviewer: 0.2.1

To start using the tool run:

```shell
./sda-cli help
```

#### Examples of common usage

- Encrypt and upload a file to the SDA in one go:

```shell

> ./sda-cli upload -config s3cmd.conf --encrypt-with-key <sda-c4gh-public-key> <unencrypted_file_to_upload>

Reviewer 2: ```suggestion
./sda-cli -config s3cmd.conf upload --encrypt-with-key <sda-c4gh-public-key> <unencrypted_file_to_upload>
```

So as for other `sda-cli` commands

```

- Encrypt and upload a whole folder recursively to a specified path, which can be different from the source, in one go:

```shell
./sda-cli upload -config s3cmd.conf --encrypt-with-key <sda-c4gh-public-key> -r <folder_1_to_upload> -targetDir <upload_folder>
```

- List all uploaded files in the user's bucket recursively:

```shell
./sda-cli list -config s3cmd.conf
```

For detailed documentation on the tool's capabilities and usage please refer [here](https://github.com/NBISweden/sda-cli#usage).

> ### Downloading data

Reviewer 3: ```suggestion
#### Downloading data
```
To stay consistent with earlier sections

Reviewer 3: Well, assuming this section is under the section "The sda-cli tool". If It's a new stand alone section it makes sense as it is

Author: Yeah its a new section from "The sda-cli tool".

> Users can directly download data from the SDA via `sda-download`, for more details see the service's [api reference](https://github.com/neicnordic/sda-download/blob/main/docs/API.md). In short, given a [valid JW token](#sda-auth), `$token`,  a user can download the file with file ID, `$fileID` by issuing the following command:

Reviewer 2: ```suggestion
Users can directly download data from the SDA via `sda-download`, for more details see the service's [api reference](https://github.com/neicnordic/sensitive-data-archive/blob/main/sda-download/api/api.md). In short, given a [valid JW token](#sda-auth), `$token`,  a user can download the file with file ID, `$fileID` by issuing the following command:
```

Also, the documentation for sda-download is outdated. I created this issue https://github.com/neicnordic/sensitive-data-archive/issues/1996 to address it.

Author: Updated link, and for now the text to mention we only support delivering files encrypted with a key provided by the client

```shell
curl --cacert <path-to-certificate-file> -H "Authorization: Bearer $token" https://<sda-download_DOMAIN_NAME>/files/$fileID -o <output-filename>
```

where for example `sda-download_DOMAIN_NAME` can be `login.gdi.nbis.se` or `localhost:8443` depending on the deployment. In the case of a local deployment, the certificate file can be obtained by running:

```shell
docker cp download:/shared/cert/ca.crt .
```

The `fileID` is a unique file identifier that can be obtained by calls to `sda-download`'s `/datasets` endpoint. For details and a concrete example on how to use `sda-download` with demo data please see [here](./README.md#get-token-or-downloading-data).

### Data access permissions

In order for a user to access a file, permission to access the dataset that the file belongs to is needed. This is granted through [REMS](https://github.com/CSCfi/rems) in the form of `GA4GH` visas. For details see [starter-kit documentation on REMS](https://github.com/GenomicDataInfrastructure/starter-kit-rems) and the links therein.

## How to perform common admin tasks

> ### The sda-admin tool

Reviewer 3: ```suggestion
### The sda-admin tool
```

This reads a bit clearer to me and feels consistent with the earlier sections

```shell
token=$(curl -s -k https://localhost:8080/tokens | jq -r '.[0]')
```

This token is created upon deployment. See `scripts/make_credentials.sh` for more details. Note that the API returns a list of tokens where the first element is the token of interest, and the rest are tokens for [testing  `sda-download`](https://github.com/neicnordic/sda-download/blob/main/dev_utils/README.md#get-a-token).

#### List datasets

```shell
curl -s -H "Authorization: Bearer $token" http://localhost:8443/metadata/datasets | jq .
```

#### List files in a dataset

```shell
datasetID=$(curl -s -H "Authorization: Bearer $token" http://localhost:8443/metadata/datasets | jq -r .'[0]')
curl -s -H "Authorization: Bearer $token" "http://localhost:8443/metadata/datasets/$datasetID/files" | jq .
```

#### Download a specific encrypted file

> The `sda-download` service offers multiple methods for downloading files through the API, with options for both encrypted and unencrypted results. Below, you will find an example illustrating each of these methods.

Reviewer 2: We need to discuss if downloading `unencrypted data` should be supported.
